drivers/gps/nrf9160_gps/nrf9160_gps.c

```diff
@@ -191,80 +172,108 @@ static void notify_event(const struct device *dev, struct gps_event *evt)
-			errno);
-		return -EIO;
-	}
+	struct nrf_modem_gnss_nmea_data_frame *data;
 
-	return 0;
+	while (k_msgq_get(&nmea_q, &data, K_NO_WAIT) == 0) {
+		k_free(data);
+	}
 }
 
-static void cancel_works(struct gps_drv_data *drv_data)
+static void nrf91_gnss_event_handler(int event)
 {
-	k_delayed_work_cancel(&drv_data->timeout_work);
-	k_delayed_work_cancel(&drv_data->blocked_work);
-}
+	struct gps_drv_data *drv_data = dev_hook->data;
+	struct nrf_modem_gnss_nmea_data_frame *nmea_data;
 
-static int gps_priority_set(struct gps_drv_data *drv_data, bool enable)
-{
-	int retval;
-	nrf_gnss_delete_mask_t delete_mask = 0;
+	switch (event) {
+	case NRF_MODEM_GNSS_EVT_PVT:
+		atomic_clear(&drv_data->has_fix);
 
-	if (enable) {
-		retval = nrf_setsockopt(drv_data->socket,
-					NRF_SOL_GNSS,
-					NRF_SO_GNSS_ENABLE_PRIORITY, NULL, 0);
-		if (retval != 0) {
-			return -EIO;
+		if (atomic_get(&drv_data->blocked)) {
+			/* Avoid spamming the logs and app. */
+			goto out;
 		}
 
-		LOG_DBG("GPS priority enabled");
-	} else {
-		retval = nrf_setsockopt(drv_data->socket,
-					NRF_SOL_GNSS,
-					NRF_SO_GNSS_DISABLE_PRIORITY, NULL, 0);
-		if (retval != 0) {
-			return -EIO;
+		k_work_schedule(&drv_data->pvt_ntf_work, K_NO_WAIT);
+		break;
+
+	case NRF_MODEM_GNSS_EVT_FIX:
+		atomic_set(&drv_data->has_fix, true);
+		drv_data->fix_timestamp = k_uptime_get();
+		k_work_submit(&drv_data->fix_ntf_work);
+		break;
+
+	case NRF_MODEM_GNSS_EVT_NMEA:
+		nmea_data = k_malloc(sizeof(struct nrf_modem_gnss_nmea_data_frame));
+
+		if (!nmea_data) {
+			return;
 		}
 
-		LOG_DBG("GPS priority disabled");
-	}
+		if (nrf_modem_gnss_read(
+			    nmea_data,
+			    sizeof(struct nrf_modem_gnss_nmea_data_frame),
+			    NRF_MODEM_GNSS_DATA_NMEA) != 0) {
+			k_free(nmea_data);
+			goto out;
+		}
 
-	/* The GPS has to be started again here because setting the option
-	 * NRF_SO_GNSS_ENABLE_PRIORITY or NRF_SO_GNSS_DISABLE_PRIORITY
-	 * implicitly stops the GPS.
-	 */
-	retval = nrf_setsockopt(drv_data->socket,
-				NRF_SOL_GNSS,
-				NRF_SO_GNSS_START,
-				&delete_mask,
-				sizeof(delete_mask));
-	if (retval != 0) {
-		LOG_ERR("Failed to start GPS");
-		return -EIO;
+		if (k_msgq_put(&nmea_q, &nmea_data, K_NO_WAIT) != 0) {
+			k_free(nmea_data);
+			goto out;
+		}
+
+		k_work_reschedule(&drv_data->nmea_ntf_work, K_NO_WAIT);
+
+		break;
+
+	case NRF_MODEM_GNSS_EVT_AGPS_REQ:
+#ifdef CONFIG_NRF9160_AGPS
+		nrf_modem_gnss_read(&drv_data->last_agps,
+				sizeof(drv_data->last_agps),
+				event);
+
+		k_work_reschedule(&drv_data->agps_ntf_work, K_NO_WAIT);
+#endif /* CONFIG_NRF9160_AGPS */
+		break;
+	case NRF_MODEM_GNSS_EVT_BLOCKED:
+		atomic_set(&drv_data->blocked, true);
+		k_work_reschedule(&drv_data->blocked_ntf_work, K_NO_WAIT);
+		break;
+	case NRF_MODEM_GNSS_EVT_UNBLOCKED:
+		atomic_clear(&drv_data->blocked);
+		k_work_reschedule(&drv_data->unblocked_ntf_work, K_NO_WAIT);
+		break;
+	default:
+		LOG_INF("Received unsupported nRF modem GNSS event");
```

Comment:
There will be other events already with MFW 1.3.0. At least in asset_tracker GPS driver log level seems to be set to debug, so this trace will be output every time GNSS starts or stops in periodic mode. Maybe we should remove the trace and just ignore unknown events?

Reply:
Fixed.